sync/src/snapshot/service.rs

```diff
@@ -55,16 +60,88 @@ impl ChainNotify for Service {
             .map(|hash| self.client.block_number(BlockId::Hash(*hash)).expect("Enacted block must exist"))
             .any(|number| number % self.period == 0);
         if is_checkpoint && best_number > self.period {
+            let client = self.client.clone();
             let root_dir = self.root_dir.clone();
             let period = self.period;
             spawn(move || {
-                let target = (best_number / period - 1) * period;
-                let path: PathBuf = [root_dir, target.to_string()].iter().collect();
-                if let Ok(_) = create_dir_all(path) {
-                    // FIXME: implement this
-                    // unimplemented!()
+                let number = (best_number / period - 1) * period;
+                let header = client.block_header(BlockId::Number(number)).expect("Snapshot target must exist");
+                let path: PathBuf = [root_dir, format!("{:x}", header.hash())].iter().collect();
+                match create_dir_all(&path) {
+                    Ok(_) => {}
+                    // FIXME: Handle file system errors
+                    Err(_) => return,
+                }
+                let db = client.database();
+
+                // FIXME: Handle non-existant nodes
+                let root = if let Some(node) = get_node(&db, &header.state_root()) {
+                    (header.state_root(), node)
+                } else {
+                    return
+                };
+                let children = if let Some(nodes) = children_of(&db, &root.1) {
+                    nodes
+                } else {
+                    return
+                };
+                let mut grandchildren = Vec::new();
+                for (_, value) in &children {
+                    if let Some(nodes) = children_of(&db, &value) {
+                        grandchildren.extend(nodes);
+                    } else {
+                        return
+                    }
+                }
+                match File::create(path.join("head")) {
+                    Ok(mut file) => {
+                        let mut stream = RlpStream::new();
+                        stream.begin_unbounded_list();
+                        for (key, value) in vec![root].iter().chain(&grandchildren).chain(&children) {
+                            stream.begin_list(2);
+                            stream.append(key);
+                            stream.append(value);
+                        }
+                        stream.complete_unbounded_list();
+                        match file.write(&stream.drain()) {
+                            Ok(_) => {}
+                            Err(_) => return,
+                        };
+                    }
+                    Err(_) => return,
```

Comment:
I saw many early returns in this function.
How about making it return ```Result```

Reply:
I'm not sure if `spawn` allows returning `Result` type. I'll check it again.

Reply:
https://doc.rust-lang.org/std/thread/fn.spawn.html `spawn` can return the `Result` if it the `Result` is `Send`.